omnichain/multioutput/contracts/MultiOutput.sol

```diff
@@ -74,21 +82,67 @@ contract MultiOutput is zContract, Ownable {
                 amountToTransfer += leftOver;
             }
 
-            uint256 outputAmount = SwapHelperLib._doSwap(
-                systemContract.wZetaContractAddress(),
-                systemContract.uniswapv2FactoryAddress(),
-                systemContract.uniswapv2Router02Address(),
-                zrc20,
-                amountToTransfer,
-                targetZRC20,
-                0
-            );
-            SwapHelperLib._doWithdrawal(
-                targetZRC20,
-                outputAmount,
-                BytesHelperLib.addressToBytes(recipient)
+            bytes memory recipient = abi.encodePacked(
+                BytesHelperLib.addressToBytes(evmRecipient)
             );
-            emit Withdrawal(targetZRC20, outputAmount, recipient);
+            
+            if (targetZRC20 == BITCOIN_ZRC20_ADDRESS) {
+                if (btcRecipient.length == 0) revert InvalidRecipient();
+                recipient = abi.encodePacked(btcRecipient);
+            }
+
+            _doSwapAndWithdraw(zrc20, amountToTransfer, targetZRC20, recipient);
+        }
+    }
+
+    function _doSwapAndWithdraw(
+        address zrc20,
+        uint256 amountToTransfer,
+        address targetZRC20,
+        bytes memory recipient
+    ) internal {
+        uint256 outputAmount = SwapHelperLib._doSwap(
+            systemContract.wZetaContractAddress(),
+            systemContract.uniswapv2FactoryAddress(),
+            systemContract.uniswapv2Router02Address(),
+            zrc20,
+            amountToTransfer,
+            targetZRC20,
+            0
+        );
+        SwapHelperLib._doWithdrawal(
+            targetZRC20,
+            outputAmount,
+            bytes32(recipient)
+        );
+        emit Withdrawal(targetZRC20, outputAmount, recipient);
+    }
+
+    function parseMessage(
+        uint256 chainID,
+        bytes calldata message)
+        public
+        pure
+        returns (address, bytes memory)
+    {
+        address evmRecipient;
+        bytes memory btcRecipient;
+        if (chainID == BITCOIN) {
+            evmRecipient = BytesHelperLib.bytesToAddress(message, 0);
+        } else {
+            if (message.length > 32) {
+                (address evmAddress, bytes memory btcAddress) = abi.decode(
```

Comment:
This assumes that the message only contains recipient addresses and we keep `registerDestinationToken`. I think the next step would be to actually remove `registerDestinationToken` and parse destination token addresses from the message.

## From Bitcoin

Tthe format of the message could be:

```
[hex-recipient][token1][token2]...
```

## From EVM

```
[hex-recipient][btc-recipient][token1][token2]...
```

On EVM we're not limited by the input data size, so we can keep BTC recipient even if the destination tokens are EVM only. This will make it simpler to parse.

We might have something like this:

```solidity
(address evmAddress, bytes memory btcAddress, bytes memory destinationTokens) = abi.decode(message, (address, bytes, bytes));
```

And then we parse `destinationTokens` to get addresses.

Reply:
It is possible to do this, but if the btcAddress is empty, it is not easy to distinguish in the contract, unless if btc is empty, we pass a default btc address in

Reply:
If the client doesn't need to swap to BTC they can fill BTC address field with any data, it will not be used by the contract.

The contract always parses the BTC address as the second field. If one of the destination tokens is BTC ZRC-20, it uses this parsed address as the withdrawal address for BTC.

Reply:
Basically, the second field is always reserved for BTC address whether it's used or not. The contract really shouldn't care until it reaches the withdrawal step, because it's just `bytes`.

If a user submitted ZRC-20 BTC as one of the destination tokens and failed to provide a valid BTC address, then, well, the contract will fail.

Reply:
Well, that's no problem. I'll finish it later

Reply:
@fadeev I updated the contract, please check it.

Reply:
Haven't tested it yet, but on a high-level I think it's looking pretty good!